keps/0008-node-heartbeat.md

```diff
@@ -0,0 +1,376 @@
+}
+
+type LeaderElectionSpec struct {
+  HolderIdentity       string      `json:"holderIdentity"`
+  LeaseDurationSeconds int32       `json:"leaseDurationSeconds"`
+  AcquireTime          metav1.Time `json:"acquireTime"`
+  RenewTime            metav1.Time `json:"renewTime"`
+  LeaderTransitions    int32       `json:"leaderTransitions"`
+}
+```
+
+The Spec is effectively of already existing (and thus proved) [LeaderElectionRecord][].
+That would hopefully allow us go get directly to Beta.
+
+We will use CRD (as opposed to built-in API), because:
+
+- CRDs are `the` way to created new APIs
+- even though CRDs aren't very efficient due to lack of protobuf support now:
+    - performance should be acceptable (because LeaderElection object will be small)
+      <br> TODO(wojtek-t): Working on microbenchmark to prove with data.
+    - protobuf support for CRDs is on the roadmap (though not near term)
+
+We will change Kubelet so that:
+
+1. Kubelet is periodically computing NodeStatus every 10s (at it is now), but that will
+   be independent from reporting status
+1. Kubelet is reporting NodeStatus if:
+   - there was a meaningful change in it (initially we can probably assume that every
+     change is meaningful, including e.g. images on the node)
+   - or it didn’t report it over last `node-status-update-period` seconds
+1. Kubelet creates and periodically updates its own LeaderElection object and frequency
+   of those updates is independent from NodeStatus update frequency.
+
+In the meantime, we will change `NodeController` to treat both updates of NodeStatus
+object as well as updates of the new `LeaderElection` object corresponding to a given
+node as healthiness signal from a given Kubelet. This will make it work for both old
+and new Kubelets.
+
+We should also:
+
+1. audit all other existing core controllers to verify if they also don’t require
+   similar changes in their logic ([ttl controller][] being one of the examples)
+1. change controller manager to auto-register that `LeaderElection` CRD
+1. ensure that `LeaderElection` resource is deleted when corresponding node is
+   deleted (probably via owner references)
+1. [out-of-scope] migrate all LeaderElection code to use that CRD
+
+Once all the code changes are done, we will:
+
+1. start updating `LeaderElection` object every 10s by default, at the same time
+   reducing frequency of NodeStatus updates initially to 20s by default.
+   We will reduce it further later.
+   <br> TODO: That still results in higher average QPS. It should be acceptable but
+   needs to be verified.
+1. announce that we are going to reduce frequency of NodeStatus updates further
+   and give people 1-2 releases to switch their code to use `LeaderElection`
+   object (if they relied on frequent NodeStatus changes)
+1. further reduce NodeStatus updates frequency to not less often than once per
+   1 minute.
+   We can’t stop periodically updating NodeStatus as it would be API breaking change,
```

Comment:
Just to confirm, today's `NodeStatus` update will be kept, as other components/clients still need a way to consume status of nodes, right?

While if that's the case, isn't it true that consumers may get stale status of Node? Is this expected?

Reply:
If something material changes in the node status, it will be updated by kubelet, as it is now.

If the node is deemed MIA, the node controller will update the node status and taints, as it does now.

Reply:
Then what would be the API breaking change if we stop the periodically NodeStatus update?